docs/source/developer_guides/model_merging.md

```diff
@@ -0,0 +1,136 @@
+<!--Copyright 2024 The HuggingFace Team. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+the License. You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# Model merging
+
+Training a model for each task can be costly, take up storage space, and the models aren't able to learn new information to improve their performance. Multitask learning can overcome some of these limitations by training a model to learn several tasks, but it is expensive to train and designing a dataset for it is challenging. *Model merging* offers a solution to these challenges by combining multiple pretrained models into one model, giving it the combined abilities of each individual model without any additional training.
+
+PEFT provides two methods for model merging:
+
+* [TIES](https://hf.co/papers/2306.01708) - TrIm, Elect, and Merge (TIES) is a three-step method for merging models. First, redundant parameters are trimmed, then conflicting signs are resolved into an aggregated vector, and finally the parameters whose signs are the same as the aggregate sign are averaged. This method takes into account that some values (redundant and sign disagreement) can degrade performance in the merged model.
+* [DARE](https://hf.co/papers/2311.03099) - Drop And REscale is a method that can be used to prepare for other model merging methods like TIES. It works by randomly dropping parameters according to a drop rate and rescaling the remaining parameters. This helps to reduce the number of redundant and potentially interfering parameters among multiple models.
+
+Models are merged with the [`~LoraModel.add_weighted_adapter`] method, and the specific model merging method is specified in the `combination_type` parameter. This guide will show you how to merge models with TIES and DARE.
+
+## Merge method
+
+With PEFT, merging is enabled by setting `combination_type` and `ties_density` to a value of the weights to keep from the individual models. For example, let's merge three finetuned [TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T](https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T) models: [tinyllama_lora_nobots](https://huggingface.co/smangrul/tinyllama_lora_norobots), [tinyllama_lora_sql](https://huggingface.co/smangrul/tinyllama_lora_sql), and [tinyllama_lora_adcopy](https://huggingface.co/smangrul/tinyllama_lora_adcopy).
```

Comment:
Moreover, I am not sure what is the right place to say this, given that ties do disjoint mean hence typically values >=1 for the weights works best because this preserves the correct scale. Moreover, all ones are a good place to start if it is hard to select the weight values. Adding a tip for this can encourage people to explore values >=1 which in my experience works best.

Reply:
Added a note in the TIES section to try using weight values >= 1.0 for better results.

Reply:
I'm having different experience here. In my experiments based on LoRAs for diffusers/LLms, I find that `ties` with `majority_sign_method="frequency"` instead of the default `majority_sign_method="total"` is working better. Also, the weights of the adapters <=1 based on the relative weightage that we want to give is working better. For `ties` with `majority_sign_method="total"`, weighting has to be done carefully.

Reply:
I have not tested TIES on diffusion models, but I could believe that to be true. When we were writing the TIES merging paper, we had many other methods of selecting signs based on normalized movement, frequency, etc which led to a few points improvements over TIES on average across many tasks but in the end, we did not add them because the optimal params depends on the distribution of the task vectors. So to make it simple we went with total everywhere. Maybe then its better to just set the default to weights = 1 with 20% unpruned parameters and then let the user try out what values they like most.